### PR TITLE
feat(MeshPassthrough): implementation of the new policy

### PR DESCRIPTION
**Pull request by lukidzi**
### Checklist prior to review

xref: https://github.com/kumahq/kuma/issues/10300 

* implementation of api
* validation if there are different protocols http/http2/grpc at the same port
* added e2e test
* added unit test with golden files

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+		// then
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func(g Gomega) {
+			resp, err := client.CollectFailure(
+				kubernetes.Cluster, "demo-client", "external-service.mesh-passthrough-mes.svc.cluster.local:80",
+				client.FromKubernetesPod(namespace, "demo-client"),
+			)
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(resp.Exitcode).To(Equal(56))
```
Comment (slonka) [inline test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go:104]: nitpick: we could have a constant here like `curlRecvError:=56`

```suggestion
			g.Expect(resp.Exitcode).To(Equal(curlRecvError))
```

File: test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		}, "30s", "1s").MustPassRepeatedly(3).Should(Succeed())
+	})
+
+	It("should control traffic to domains", func() {
```
Comment (slonka) [inline test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go:108]: nitpick:

```suggestion
	It("should control traffic to domains and IPs", func() {
```

File: test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  default:
+    appendMatch:
+    - type: Domain
+      value: external-service.mesh-passthrough-mes.svc.cluster.local
+      port: 80
+      protocol: http
+    - type: IP
+      value: %s
+      port: 80
+      protocol: http
```
Comment (slonka) [inline test/e2e_env/kubernetes/meshpassthrough/meshpassthrough.go:163]: we're missing CIDR but I guess it's ok to just have that tested in unit tests

File: pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/testdata/basic.clusters.golden.yaml [hunk 4]
```diff
@@ -0,0 +1,134 @@
+resources:
+- name: meshpassthrough_*.example.com_*
+  resource:
+    '@type': type.googleapis.com/envoy.config.cluster.v3.Cluster
+    altStatName: meshpassthrough___example_com__
```
Comment (slonka) [inline pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/testdata/basic.clusters.golden.yaml:5]: should we think about replacing `*` with something else than `_`? it's going to be hard to distinguish it, right?

**Review by slonka — COMMENTED** [review]
In general good, but I feel like these https://github.com/kumahq/kuma/pull/10363/files#diff-cc64881745ca77b7290fa733600b61a7b48240acd64de547c31816ed88218526R33-R141 could be yamls so it's a bit easier to read

File: pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/plugin_test.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+	resources, err := resourceSet.ListOf(typ).ToDeltaDiscoveryResponse()
+	Expect(err).ToNot(HaveOccurred())
+	actual, err := util_proto.ToYAML(resources)
+	Expect(err).ToNot(HaveOccurred())
+
+	return actual
+}
+
+var _ = Describe("MeshPassthrough", func() {
+	type sidecarTestCase struct {
```
Comment (slonka) [inline pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/plugin_test.go:41]: I think it's a leftover from `sidecar` test

File: pkg/plugins/policies/meshpassthrough/plugin/xds/order.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+	lenI := len(splitI)
+	lenJ := len(splitJ)
+
+	if lenI != lenJ {
+		return lenI > lenJ
+	}
+
+	return i < j
+}
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/plugins/policies/meshpassthrough/plugin/xds/order.go:166]: Yes, that's the case

File: pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/testdata/basic.clusters.golden.yaml (hunk 4 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/testdata/basic.clusters.golden.yaml:5]: Do you have maybe some idea? The `altStatName` is replaced building cluster/filter chain so it's not replaced by me

Comment (slonka) [reply pkg/plugins/policies/meshpassthrough/plugin/v1alpha1/testdata/basic.clusters.golden.yaml:5]: we could do something like wildcard, so it's `meshpassthrough_wildcard_example_com_wildcard`?

File: pkg/plugins/policies/core/ordered/ordered.go [hunk 7]
```diff
@@ -25,6 +25,8 @@ var Policies = []plugins.PluginName{
 	plugins.PluginName(meshtcproute_api.MeshTCPRouteResourceTypeDescriptor.KumactlArg),
 	// For other policies order isn't important at the moment
 	plugins.PluginName(meshloadbalancingstrategy_api.MeshLoadBalancingStrategyResourceTypeDescriptor.KumactlArg),
+	// has to be before MeshAccessLog so logs can access logs are added to the chains
```
Comment (slonka) [inline pkg/plugins/policies/core/ordered/ordered.go:28]: sounds a bit weird, "logs can access logs" ? shouldn't this be 

```suggestion
	// has to be before MeshAccessLog so the plugin can access log filters that are added to the filter chains
```

?

**Review by slonka — APPROVED** [review]
just one small thing, should be good to go 👍

**Pull request merged**